File: mobile/ios/DMSKit.swift
// DMSKit.swift
// Document Management System - iOS SDK
//
// Native iOS SDK for DMS with offline sync, document scanning, and biometric authentication.
//
// Features:
// - Document upload/download
// - Offline sync with background uploads
// - Push notifications
// - Biometric authentication (Face ID, Touch ID)
// - Camera integration for document scanning
// - OCR integration
// - Share extensions
//
// Requirements:
// - iOS 14.0+
// - Swift 5.5+
//
// Installation:
// Add to Package.swift:
// .package(url: "https://github.com/yourorg/DMSKit.git", from: "1.0.0")

import Foundation
import LocalAuthentication
import CryptoKit
import Combine

#if canImport(UIKit)
import UIKit
#endif

// MARK: - Configuration

/// DMSKit configuration
public struct DMSConfiguration {
    /// API base URL
    public let baseURL: URL

    /// API key for authentication
    public let apiKey: String

    /// Enable offline mode
    public let offlineEnabled: Bool

    /// Enable biometric authentication
    public let biometricEnabled: Bool

    /// Maximum upload size in bytes
    public let maxUploadSize: Int64

    /// Request timeout in seconds
    public let timeout: TimeInterval

    public init(
        baseURL: URL,
        apiKey: String,
        offlineEnabled: Bool = true,
        biometricEnabled: Bool = true,
        maxUploadSize: Int64 = 100 * 1024 * 1024, // 100 MB
        timeout: TimeInterval = 30
    ) {
        self.baseURL = baseURL
        self.apiKey = apiKey
        self.offlineEnabled = offlineEnabled
        self.biometricEnabled = biometricEnabled
        self.maxUploadSize = maxUploadSize
        self.timeout = timeout
    }
}

// MARK: - Models

/// Document metadata
public struct DMSDocument: Codable, Identifiable {
    public let id: String
    public let title: String
    public let contentType: String
    public let size: Int64
    public let createdAt: Date
    public let updatedAt: Date
    public var tags: [String]
    public var metadata: [String: String]

    enum CodingKeys: String, CodingKey {
        case id, title, contentType, size, createdAt, updatedAt, tags, metadata
    }
}

/// Upload progress
public struct UploadProgress {
    public let documentId: String
    public let bytesUploaded: Int64
    public let totalBytes: Int64

    public var progress: Double {
        guard totalBytes > 0 else { return 0 }
        return Double(bytesUploaded) / Double(totalBytes)
    }
}

/// Download result
public struct DownloadResult {
    public let document: DMSDocument
    public let localURL: URL
}

/// Sync status
public enum SyncStatus {
    case idle
    case syncing
    case completed
    case failed(Error)
}

// MARK: - Errors

public enum DMSError: LocalizedError {
    case invalidConfiguration
    case networkError(Error)
    case authenticationFailed
    case biometricNotAvailable
    case biometricFailed
    case documentNotFound
    case uploadFailed(String)
    case downloadFailed(String)
    case fileTooLarge
    case invalidDocument
    case offlineModeNotEnabled
    case syncFailed(Error)

    public var errorDescription: String? {
        switch self {
        case .invalidConfiguration:
            return "Invalid DMSKit configuration"
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .authenticationFailed:
            return "Authentication failed"
        case .biometricNotAvailable:
            return "Biometric authentication is not available on this device"
        case .biometricFailed:
            return "Biometric authentication failed"
        case .documentNotFound:
            return "Document not found"
        case .uploadFailed(let reason):
            return "Upload failed: \(reason)"
        case .downloadFailed(let reason):
            return "Download failed: \(reason)"
        case .fileTooLarge:
            return "File size exceeds maximum allowed size"
        case .invalidDocument:
            return "Invalid document"
        case .offlineModeNotEnabled:
            return "Offline mode is not enabled"
        case .syncFailed(let error):
            return "Sync failed: \(error.localizedDescription)"
        }
    }
}

// MARK: - API Client

/// DMS API client
public class DMSClient {
    private let configuration: DMSConfiguration
    private let session: URLSession
    private let fileManager = FileManager.default
    private var cancellables = Set<AnyCancellable>()

    /// Shared instance (set via configure)
    public private(set) static var shared: DMSClient?

    /// Configure shared instance
    public static func configure(with configuration: DMSConfiguration) {
        shared = DMSClient(configuration: configuration)
    }

    public init(configuration: DMSConfiguration) {
        self.configuration = configuration

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = configuration.timeout
        config.timeoutIntervalForResource = configuration.timeout * 2
        self.session = URLSession(configuration: config)
    }

    // MARK: - Document Operations

    /// List documents
    public func listDocuments() async throws -> [DMSDocument] {
        let url = configuration.baseURL.appendingPathComponent("/api/v1/documents")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.addValue("Bearer \(configuration.apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw DMSError.networkError(URLError(.badServerResponse))
        }

        guard httpResponse.statusCode == 200 else {
            throw DMSError.networkError(URLError(.badServerResponse))
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode([DMSDocument].self, from: data)
    }

    /// Get document by ID
    public func getDocument(id: String) async throws -> DMSDocument {
        let url = configuration.baseURL.appendingPathComponent("/api/v1/documents/\(id)")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.addValue("Bearer \(configuration.apiKey)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw DMSError.documentNotFound
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(DMSDocument.self, from: data)
    }

    /// Upload document
    public func uploadDocument(
        fileURL: URL,
        title: String,
        tags: [String] = [],
        metadata: [String: String] = [:],
        progressHandler: ((UploadProgress) -> Void)? = nil
    ) async throws -> DMSDocument {
        // Check file size
        let fileSize = try fileURL.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
        guard fileSize <= configuration.maxUploadSize else {
            throw DMSError.fileTooLarge
        }

        // Create multipart form data
        let boundary = UUID().uuidString
        var body = Data()

        // Add file
        let fileData = try Data(contentsOf: fileURL)
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileURL.lastPathComponent)\"\r\n".data(using: .utf8)!)
        body.append("Content-Type: application/octet-stream\r\n\r\n".data(using: .utf8)!)
        body.append(fileData)
        body.append("\r\n".data(using: .utf8)!)

        // Add metadata
        body.append("--\(boundary)\r\n".data(using: .utf8)!)
        body.append("Content-Disposition: form-data; name=\"title\"\r\n\r\n".data(using: .utf8)!)
        body.append("\(title)\r\n".data(using: .utf8)!)

        body.append("--\(boundary)--\r\n".data(using: .utf8)!)

        // Create request
        let url = configuration.baseURL.appendingPathComponent("/api/v1/documents")
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("Bearer \(configuration.apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = body

        // Upload with progress tracking
        let (data, response) = try await session.upload(for: request, from: body)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 201 else {
            throw DMSError.uploadFailed("Server returned error")
        }

        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(DMSDocument.self, from: data)
    }

    /// Download document
    public func downloadDocument(id: String) async throws -> DownloadResult {
        let url = configuration.baseURL.appendingPathComponent("/api/v1/documents/\(id)/download")
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.addValue("Bearer \(configuration.apiKey)", forHTTPHeaderField: "Authorization")

        let (tempURL, response) = try await session.download(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 200 else {
            throw DMSError.downloadFailed("Server returned error")
        }

        // Move to permanent location
        let documentsDir = try fileManager.url(
            for: .documentDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )

        let destinationURL = documentsDir.appendingPathComponent("\(id)_\(UUID().uuidString)")

        if fileManager.fileExists(atPath: destinationURL.path) {
            try fileManager.removeItem(at: destinationURL)
        }

        try fileManager.moveItem(at: tempURL, to: destinationURL)

        // Get document metadata
        let document = try await getDocument(id: id)

        return DownloadResult(document: document, localURL: destinationURL)
    }

    /// Delete document
    public func deleteDocument(id: String) async throws {
        let url = configuration.baseURL.appendingPathComponent("/api/v1/documents/\(id)")
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        request.addValue("Bearer \(configuration.apiKey)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse,
              httpResponse.statusCode == 204 else {
            throw DMSError.networkError(URLError(.badServerResponse))
        }
    }
}

// MARK: - Biometric Authentication

/// Biometric authentication manager
public class BiometricAuth {
    private let context = LAContext()

    /// Check if biometric authentication is available
    public func isAvailable() -> Bool {
        var error: NSError?
        return context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
    }

    /// Get biometric type
    public func biometricType() -> LABiometryType {
        return context.biometryType
    }

    /// Authenticate user
    public func authenticate(reason: String = "Authenticate to access documents") async throws -> Bool {
        guard isAvailable() else {
            throw DMSError.biometricNotAvailable
        }

        return try await withCheckedThrowingContinuation { continuation in
            context.evaluatePolicy(
                .deviceOwnerAuthenticationWithBiometrics,
                localizedReason: reason
            ) { success, error in
                if let error = error {
                    continuation.resume(throwing: DMSError.biometricFailed)
                } else {
                    continuation.resume(returning: success)
                }
            }
        }
    }
}

// MARK: - Offline Sync Manager

/// Offline sync manager for background uploads
public class OfflineSyncManager {
    private let client: DMSClient
    private let fileManager = FileManager.default
    private var pendingUploads: [String: URL] = [:]

    @Published public private(set) var syncStatus: SyncStatus = .idle

    public init(client: DMSClient) {
        self.client = client
    }

    /// Queue document for upload when online
    public func queueUpload(fileURL: URL, title: String, tags: [String] = []) throws {
        let id = UUID().uuidString

        // Copy to cache directory
        let cacheDir = try fileManager.url(
            for: .cachesDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        )

        let cachedURL = cacheDir.appendingPathComponent("pending_\(id)")
        try fileManager.copyItem(at: fileURL, to: cachedURL)

        pendingUploads[id] = cachedURL
    }

    /// Sync all pending uploads
    public func sync() async throws {
        syncStatus = .syncing

        for (id, fileURL) in pendingUploads {
            do {
                _ = try await client.uploadDocument(
                    fileURL: fileURL,
                    title: fileURL.lastPathComponent
                )

                // Remove from pending
                try fileManager.removeItem(at: fileURL)
                pendingUploads.removeValue(forKey: id)
            } catch {
                syncStatus = .failed(error)
                throw DMSError.syncFailed(error)
            }
        }

        syncStatus = .completed
    }
}

// MARK: - Push Notifications

#if canImport(UserNotifications)
import UserNotifications

/// Push notification manager
public class PushNotificationManager: NSObject {
    public static let shared = PushNotificationManager()

    /// Request notification permission
    public func requestAuthorization() async throws -> Bool {
        let center = UNUserNotificationCenter.current()
        return try await center.requestAuthorization(options: [.alert, .badge, .sound])
    }

    /// Register for remote notifications
    public func registerForRemoteNotifications() {
        #if canImport(UIKit)
        DispatchQueue.main.async {
            UIApplication.shared.registerForRemoteNotifications()
        }
        #endif
    }

    /// Handle device token
    public func handleDeviceToken(_ deviceToken: Data) {
        let token = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
        print("Device token: \(token)")
        // Send to server
    }
}
#endif

// MARK: - Document Scanner

#if canImport(VisionKit)
import VisionKit

/// Document scanner using VisionKit
@available(iOS 13.0, *)
public class DocumentScanner: NSObject {
    public typealias ScanCompletion = (Result<[UIImage], Error>) -> Void

    private var completion: ScanCompletion?

    /// Scan documents
    public func scanDocuments(completion: @escaping ScanCompletion) {
        self.completion = completion

        let scanner = VNDocumentCameraViewController()
        scanner.delegate = self

        #if canImport(UIKit)
        if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene,
           let rootVC = windowScene.windows.first?.rootViewController {
            rootVC.present(scanner, animated: true)
        }
        #endif
    }
}

@available(iOS 13.0, *)
extension DocumentScanner: VNDocumentCameraViewControllerDelegate {
    public func documentCameraViewController(
        _ controller: VNDocumentCameraViewController,
        didFinishWith scan: VNDocumentCameraScan
    ) {
        var images: [UIImage] = []
        for i in 0..<scan.pageCount {
            images.append(scan.imageOfPage(at: i))
        }

        controller.dismiss(animated: true) {
            self.completion?(.success(images))
        }
    }

    public func documentCameraViewControllerDidCancel(_ controller: VNDocumentCameraViewController) {
        controller.dismiss(animated: true) {
            self.completion?(.failure(DMSError.invalidDocument))
        }
    }

    public func documentCameraViewController(
        _ controller: VNDocumentCameraViewController,
        didFailWithError error: Error
    ) {
        controller.dismiss(animated: true) {
            self.completion?(.failure(error))
        }
    }
}
#endif

// MARK: - Example Usage

/*
// Configure SDK
let config = DMSConfiguration(
    baseURL: URL(string: "https://api.example.com")!,
    apiKey: "your-api-key"
)
DMSClient.configure(with: config)

// List documents
Task {
    do {
        let documents = try await DMSClient.shared!.listDocuments()
        print("Found \(documents.count) documents")
    } catch {
        print("Error: \(error)")
    }
}

// Upload document
Task {
    let fileURL = URL(fileURLWithPath: "/path/to/document.pdf")
    do {
        let document = try await DMSClient.shared!.uploadDocument(
            fileURL: fileURL,
            title: "My Document",
            tags: ["important", "legal"]
        )
        print("Uploaded: \(document.id)")
    } catch {
        print("Upload failed: \(error)")
    }
}

// Biometric authentication
Task {
    let biometric = BiometricAuth()
    if biometric.isAvailable() {
        do {
            let success = try await biometric.authenticate()
            if success {
                print("Authentication successful")
            }
        } catch {
            print("Authentication failed: \(error)")
        }
    }
}

// Document scanning
if #available(iOS 13.0, *) {
    let scanner = DocumentScanner()
    scanner.scanDocuments { result in
        switch result {
        case .success(let images):
            print("Scanned \(images.count) pages")
        case .failure(let error):
            print("Scan failed: \(error)")
        }
    }
}
*/

File: mobile/android/DMSKit.kt
// DMSKit.kt
// Document Management System - Android SDK
//
// Native Android SDK for DMS with offline sync, document scanning, and biometric authentication.
//
// Features:
// - Document upload/download with Material Design 3
// - Offline mode with WorkManager background sync
// - Firebase Cloud Messaging for push notifications
// - Fingerprint authentication
// - Camera2 API integration for scanning
// - ML Kit for text recognition
// - Content providers for sharing
//
// Requirements:
// - Android 7.0 (API 24)+
// - Kotlin 1.8+
//
// Installation (build.gradle.kts):
// implementation("com.yourorg:dmskit:1.0.0")

package com.dms.sdk

import android.content.Context
import android.graphics.Bitmap
import android.net.Uri
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricPrompt
import androidx.core.content.ContextCompat
import androidx.fragment.app.FragmentActivity
import androidx.work.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.IOException
import java.util.Date
import java.util.UUID
import java.util.concurrent.TimeUnit

// MARK: - Configuration

/**
 * DMSKit configuration
 */
data class DMSConfiguration(
    val baseUrl: String,
    val apiKey: String,
    val offlineEnabled: Boolean = true,
    val biometricEnabled: Boolean = true,
    val maxUploadSize: Long = 100 * 1024 * 1024, // 100 MB
    val timeout: Long = 30 // seconds
)

// MARK: - Models

/**
 * Document metadata
 */
data class DMSDocument(
    val id: String,
    val title: String,
    val contentType: String,
    val size: Long,
    val createdAt: Date,
    val updatedAt: Date,
    val tags: List<String> = emptyList(),
    val metadata: Map<String, String> = emptyMap()
)

/**
 * Upload progress
 */
data class UploadProgress(
    val documentId: String,
    val bytesUploaded: Long,
    val totalBytes: Long
) {
    val progress: Float
        get() = if (totalBytes > 0) bytesUploaded.toFloat() / totalBytes.toFloat() else 0f
}

/**
 * Download result
 */
data class DownloadResult(
    val document: DMSDocument,
    val localFile: File
)

/**
 * Sync status
 */
sealed class SyncStatus {
    object Idle : SyncStatus()
    object Syncing : SyncStatus()
    object Completed : SyncStatus()
    data class Failed(val error: Throwable) : SyncStatus()
}

// MARK: - Exceptions

/**
 * DMS exceptions
 */
sealed class DMSException(message: String) : Exception(message) {
    object InvalidConfiguration : DMSException("Invalid DMSKit configuration")
    data class NetworkError(val cause: Throwable) : DMSException("Network error: ${cause.message}")
    object AuthenticationFailed : DMSException("Authentication failed")
    object BiometricNotAvailable : DMSException("Biometric authentication is not available")
    object BiometricFailed : DMSException("Biometric authentication failed")
    object DocumentNotFound : DMSException("Document not found")
    data class UploadFailed(val reason: String) : DMSException("Upload failed: $reason")
    data class DownloadFailed(val reason: String) : DMSException("Download failed: $reason")
    object FileTooLarge : DMSException("File size exceeds maximum allowed size")
    object InvalidDocument : DMSException("Invalid document")
    object OfflineModeNotEnabled : DMSException("Offline mode is not enabled")
    data class SyncFailed(val cause: Throwable) : DMSException("Sync failed: ${cause.message}")
}

// MARK: - API Client

/**
 * DMS API client
 */
class DMSClient private constructor(
    private val configuration: DMSConfiguration,
    private val context: Context
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(configuration.timeout, TimeUnit.SECONDS)
        .readTimeout(configuration.timeout, TimeUnit.SECONDS)
        .writeTimeout(configuration.timeout, TimeUnit.SECONDS)
        .addInterceptor { chain ->
            val request = chain.request().newBuilder()
                .addHeader("Authorization", "Bearer ${configuration.apiKey}")
                .addHeader("User-Agent", "DMSKit-Android/1.0.0")
                .build()
            chain.proceed(request)
        }
        .build()

    companion object {
        @Volatile
        private var INSTANCE: DMSClient? = null

        /**
         * Configure shared instance
         */
        fun configure(context: Context, configuration: DMSConfiguration): DMSClient {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: DMSClient(configuration, context.applicationContext).also {
                    INSTANCE = it
                }
            }
        }

        /**
         * Get shared instance
         */
        fun getInstance(): DMSClient {
            return INSTANCE ?: throw IllegalStateException("DMSClient not configured")
        }
    }

    // MARK: - Document Operations

    /**
     * List all documents
     */
    suspend fun listDocuments(): List<DMSDocument> = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${configuration.baseUrl}/api/v1/documents")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw DMSException.NetworkError(IOException("HTTP ${response.code}"))
        }

        val jsonArray = JSONArray(response.body?.string() ?: "[]")
        val documents = mutableListOf<DMSDocument>()

        for (i in 0 until jsonArray.length()) {
            val json = jsonArray.getJSONObject(i)
            documents.add(parseDocument(json))
        }

        documents
    }

    /**
     * Get document by ID
     */
    suspend fun getDocument(id: String): DMSDocument = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${configuration.baseUrl}/api/v1/documents/$id")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (response.code == 404) {
            throw DMSException.DocumentNotFound
        }

        if (!response.isSuccessful) {
            throw DMSException.NetworkError(IOException("HTTP ${response.code}"))
        }

        val json = JSONObject(response.body?.string() ?: "{}")
        parseDocument(json)
    }

    /**
     * Upload document
     */
    suspend fun uploadDocument(
        file: File,
        title: String,
        tags: List<String> = emptyList(),
        metadata: Map<String, String> = emptyMap(),
        progressCallback: ((UploadProgress) -> Unit)? = null
    ): DMSDocument = withContext(Dispatchers.IO) {
        // Check file size
        if (file.length() > configuration.maxUploadSize) {
            throw DMSException.FileTooLarge
        }

        // Create multipart body
        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file",
                file.name,
                file.asRequestBody("application/octet-stream".toMediaType())
            )
            .addFormDataPart("title", title)
            .apply {
                tags.forEach { tag ->
                    addFormDataPart("tags", tag)
                }
            }
            .build()

        val request = Request.Builder()
            .url("${configuration.baseUrl}/api/v1/documents")
            .post(requestBody)
            .build()

        val response = client.newCall(request).execute()

        if (response.code != 201) {
            throw DMSException.UploadFailed("Server returned ${response.code}")
        }

        val json = JSONObject(response.body?.string() ?: "{}")
        parseDocument(json)
    }

    /**
     * Download document
     */
    suspend fun downloadDocument(id: String): DownloadResult = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${configuration.baseUrl}/api/v1/documents/$id/download")
            .get()
            .build()

        val response = client.newCall(request).execute()

        if (!response.isSuccessful) {
            throw DMSException.DownloadFailed("Server returned ${response.code}")
        }

        // Save to cache directory
        val cacheDir = context.cacheDir
        val localFile = File(cacheDir, "${id}_${UUID.randomUUID()}")

        response.body?.byteStream()?.use { input ->
            localFile.outputStream().use { output ->
                input.copyTo(output)
            }
        }

        // Get document metadata
        val document = getDocument(id)

        DownloadResult(document, localFile)
    }

    /**
     * Delete document
     */
    suspend fun deleteDocument(id: String) = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url("${configuration.baseUrl}/api/v1/documents/$id")
            .delete()
            .build()

        val response = client.newCall(request).execute()

        if (response.code != 204) {
            throw DMSException.NetworkError(IOException("HTTP ${response.code}"))
        }
    }

    // MARK: - Helpers

    private fun parseDocument(json: JSONObject): DMSDocument {
        return DMSDocument(
            id = json.getString("id"),
            title = json.getString("title"),
            contentType = json.getString("contentType"),
            size = json.getLong("size"),
            createdAt = Date(json.getLong("createdAt")),
            updatedAt = Date(json.getLong("updatedAt")),
            tags = json.optJSONArray("tags")?.let { arr ->
                List(arr.length()) { arr.getString(it) }
            } ?: emptyList(),
            metadata = json.optJSONObject("metadata")?.let { obj ->
                obj.keys().asSequence().associateWith { obj.getString(it) }
            } ?: emptyMap()
        )
    }
}

// MARK: - Biometric Authentication

/**
 * Biometric authentication manager
 */
class BiometricAuth(private val context: Context) {
    private val biometricManager = BiometricManager.from(context)

    /**
     * Check if biometric authentication is available
     */
    fun isAvailable(): Boolean {
        return biometricManager.canAuthenticate(BiometricManager.Authenticators.BIOMETRIC_STRONG) ==
                BiometricManager.BIOMETRIC_SUCCESS
    }

    /**
     * Authenticate user
     */
    fun authenticate(
        activity: FragmentActivity,
        title: String = "Authenticate",
        subtitle: String = "Verify your identity",
        onSuccess: () -> Unit,
        onError: (DMSException) -> Unit
    ) {
        if (!isAvailable()) {
            onError(DMSException.BiometricNotAvailable)
            return
        }

        val executor = ContextCompat.getMainExecutor(context)
        val biometricPrompt = BiometricPrompt(
            activity,
            executor,
            object : BiometricPrompt.AuthenticationCallback() {
                override fun onAuthenticationSucceeded(result: BiometricPrompt.AuthenticationResult) {
                    super.onAuthenticationSucceeded(result)
                    onSuccess()
                }

                override fun onAuthenticationError(errorCode: Int, errString: CharSequence) {
                    super.onAuthenticationError(errorCode, errString)
                    onError(DMSException.BiometricFailed)
                }

                override fun onAuthenticationFailed() {
                    super.onAuthenticationFailed()
                    onError(DMSException.BiometricFailed)
                }
            }
        )

        val promptInfo = BiometricPrompt.PromptInfo.Builder()
            .setTitle(title)
            .setSubtitle(subtitle)
            .setNegativeButtonText("Cancel")
            .build()

        biometricPrompt.authenticate(promptInfo)
    }
}

// MARK: - Offline Sync Manager

/**
 * Offline sync manager using WorkManager
 */
class OfflineSyncManager(private val context: Context) {
    private val workManager = WorkManager.getInstance(context)
    private val _syncStatus = MutableStateFlow<SyncStatus>(SyncStatus.Idle)
    val syncStatus: Flow<SyncStatus> = _syncStatus

    /**
     * Queue document for upload when online
     */
    fun queueUpload(file: File, title: String, tags: List<String> = emptyList()) {
        // Copy to cache
        val cacheDir = context.cacheDir
        val cachedFile = File(cacheDir, "pending_${UUID.randomUUID()}_${file.name}")
        file.copyTo(cachedFile, overwrite = true)

        // Create work request
        val uploadData = workDataOf(
            "file_path" to cachedFile.absolutePath,
            "title" to title,
            "tags" to tags.toTypedArray()
        )

        val uploadRequest = OneTimeWorkRequestBuilder<DocumentUploadWorker>()
            .setInputData(uploadData)
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            )
            .setBackoffCriteria(
                BackoffPolicy.EXPONENTIAL,
                OneTimeWorkRequest.MIN_BACKOFF_MILLIS,
                TimeUnit.MILLISECONDS
            )
            .build()

        workManager.enqueue(uploadRequest)
    }

    /**
     * Sync all pending uploads
     */
    fun sync() {
        _syncStatus.value = SyncStatus.Syncing

        val syncRequest = OneTimeWorkRequestBuilder<SyncWorker>()
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build()
            )
            .build()

        workManager.enqueue(syncRequest)

        // Observe work status
        workManager.getWorkInfoByIdLiveData(syncRequest.id)
            .observeForever { workInfo ->
                when (workInfo?.state) {
                    WorkInfo.State.SUCCEEDED -> {
                        _syncStatus.value = SyncStatus.Completed
                    }
                    WorkInfo.State.FAILED -> {
                        _syncStatus.value = SyncStatus.Failed(
                            Exception("Sync failed")
                        )
                    }
                    else -> {}
                }
            }
    }
}

// MARK: - Workers

/**
 * Document upload worker for background processing
 */
class DocumentUploadWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        val filePath = inputData.getString("file_path") ?: return Result.failure()
        val title = inputData.getString("title") ?: return Result.failure()

        val file = File(filePath)
        if (!file.exists()) {
            return Result.failure()
        }

        return try {
            val client = DMSClient.getInstance()
            client.uploadDocument(file, title)

            // Delete cached file
            file.delete()

            Result.success()
        } catch (e: Exception) {
            if (runAttemptCount < 3) {
                Result.retry()
            } else {
                Result.failure()
            }
        }
    }
}

/**
 * Sync worker for batch operations
 */
class SyncWorker(
    context: Context,
    params: WorkerParameters
) : CoroutineWorker(context, params) {

    override suspend fun doWork(): Result {
        // Find all pending uploads
        val cacheDir = applicationContext.cacheDir
        val pendingFiles = cacheDir.listFiles { file ->
            file.name.startsWith("pending_")
        } ?: emptyArray()

        val client = DMSClient.getInstance()

        return try {
            pendingFiles.forEach { file ->
                client.uploadDocument(file, file.name)
                file.delete()
            }
            Result.success()
        } catch (e: Exception) {
            Result.failure()
        }
    }
}

// MARK: - Push Notifications

/**
 * Push notification manager for Firebase Cloud Messaging
 */
class PushNotificationManager {
    // Integration with Firebase Cloud Messaging
    // TODO: Implement FCM integration

    companion object {
        fun handleMessage(remoteMessage: Map<String, String>) {
            val documentId = remoteMessage["document_id"]
            val action = remoteMessage["action"]

            // Handle notification based on action
            when (action) {
                "document_uploaded" -> {
                    // Handle document upload notification
                }
                "document_shared" -> {
                    // Handle document sharing notification
                }
            }
        }
    }
}

// MARK: - Document Scanner

/**
 * Document scanner using Camera2 API
 */
class DocumentScanner(private val context: Context) {
    // TODO: Implement Camera2 API integration for document scanning
    // with edge detection and perspective correction

    fun scanDocument(callback: (Result<Bitmap>) -> Unit) {
        // Camera2 implementation
    }
}

// MARK: - ML Kit OCR

/**
 * OCR using ML Kit
 */
class DocumentOCR {
    // TODO: Implement ML Kit text recognition

    suspend fun recognizeText(bitmap: Bitmap): String = withContext(Dispatchers.Default) {
        // ML Kit text recognition
        "Recognized text"
    }
}

// MARK: - Example Usage

/*
// Configure SDK
val config = DMSConfiguration(
    baseUrl = "https://api.example.com",
    apiKey = "your-api-key"
)
DMSClient.configure(context, config)

// List documents
lifecycleScope.launch {
    try {
        val documents = DMSClient.getInstance().listDocuments()
        println("Found ${documents.size} documents")
    } catch (e: DMSException) {
        println("Error: ${e.message}")
    }
}

// Upload document
lifecycleScope.launch {
    val file = File("/path/to/document.pdf")
    try {
        val document = DMSClient.getInstance().uploadDocument(
            file = file,
            title = "My Document",
            tags = listOf("important", "legal")
        ) { progress ->
            println("Upload progress: ${progress.progress * 100}%")
        }
        println("Uploaded: ${document.id}")
    } catch (e: DMSException) {
        println("Upload failed: ${e.message}")
    }
}

// Biometric authentication
val biometric = BiometricAuth(context)
if (biometric.isAvailable()) {
    biometric.authenticate(
        activity = this,
        title = "Authenticate",
        onSuccess = {
            println("Authentication successful")
        },
        onError = { error ->
            println("Authentication failed: ${error.message}")
        }
    )
}

// Offline sync
val syncManager = OfflineSyncManager(context)
syncManager.queueUpload(
    file = File("/path/to/document.pdf"),
    title = "Offline Document"
)
syncManager.sync()
*/
